crypto/trace.c

```diff
@@ -211,7 +218,7 @@ static int trace_detach_cb(int category, int type, const void *data)
     return 1;
 }
 
```

Comment:
It looks like all the type arguments are internal, if they are we’d be safe passing them as their enum instead of an int. Less checking for the compiler. Definitely not essential and I could easily have missed something.

Reply:
Makes sense, fixup to be done...

Reply:
Maybe I am missing something, but if the C compiler allows passing a value of `4711` as an enum parameter argument (where the enum has only three possible values) without even issuing a warning, then I really don't see how using an `enum` instead of an `int` would increase type safety. In this case I would leave everything as it is.

(any good explanations by the experts, e.g. @bernd-edlinger?)

```diff
index 613664c3be..c908853ae0 100644
--- a/crypto/trace.c
+++ b/crypto/trace.c
@@ -170,13 +170,13 @@ static struct {
 
 #ifndef OPENSSL_NO_TRACE
 
-enum {
+enum Type {
     CHANNEL,
     PREFIX,
     SUFFIX
 };
 
-static int trace_attach_cb(int category, int type, const void *data)
+static int trace_attach_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
@@ -197,7 +197,7 @@ static int trace_attach_cb(int category, int type, const void *data)
     return 1;
 }
 
-static int trace_detach_cb(int category, int type, const void *data)
+static int trace_detach_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
@@ -218,10 +218,10 @@ static int trace_detach_cb(int category, int type, const void *data)
     return 1;
 }
 
-static int set_trace_data(int category, int type, BIO **channel,
+static int set_trace_data(int category, enum Type type, BIO **channel,
                           const char **prefix, const char **suffix,
-                          int (*attach_cb)(int, int, const void *),
-                          int (*detach_cb)(int, int, const void *))
+                          int (*attach_cb)(int, enum Type, const void *),
+                          int (*detach_cb)(int, enum Type, const void *))
 {
     BIO *curr_channel = trace_channels[category].bio;
     char *curr_prefix = trace_channels[category].prefix;
@@ -315,7 +315,7 @@ void ossl_trace_cleanup(void)
         /* We force the TRACE category to be treated last */
         if (category == OSSL_TRACE_CATEGORY_TRACE)
             continue;
-        set_trace_data(category, 0, &channel, &prefix, &suffix,
+        set_trace_data(category, 4711, &channel, &prefix, &suffix,
                        trace_attach_cb, trace_detach_cb);
     }
     set_trace_data(OSSL_TRACE_CATEGORY_TRACE, 0, &channel,
@@ -337,7 +337,7 @@ int OSSL_trace_set_channel(int category, BIO *channel)
 }
 
 #ifndef OPENSSL_NO_TRACE
-static int trace_attach_w_callback_cb(int category, int type, const void *data)
+static int trace_attach_w_callback_cb(int category, enum Type type, const void *data)
 {
     switch (type) {
     case CHANNEL:
```

Reply:
if you assign a value 4711 in an enum {a,b,c};
you get undefined behaviour, you should get an undefined sanitizer warning.
There are targets where this enum with range 0-2 occupy only 1 byte.

Reply:
The only disadvantage I could imagine, if you use an enum as a callback parameter,
is if you later want to add an enum value that makes sizeof(enum) be different, it will be
an ABI change.
i.e. 
enum { a, b, c }; may be 1-4 byte integer dependent on the target (-fshort-enums in gcc makes it a byte).

and if you later add
enum { a, b, c, d=0x80000000 }; it will be a unsigned 4-byte integer

Reply:
look at test/sanitytest which tries to catch anything where it's not ```sizeof(int)```

Having said that, using enum's in an API is a bad idea anyway.

Reply:
> Having said that, using enum's in an API is a bad idea anyway.

That was my final conclusion, too. Marking this discussion resolved.